### PR TITLE
Add Marvell device diagnostics data collection.

### PR DESCRIPTION
**Pull request by vikasmrvl**
Initial submit.
Changes to collect diagnostics data for Marvell devices.

### CODE REVIEW

File: Get-NetView.psm1 [hunk 1]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+        {
+            Exception e = Marshal.GetExceptionForHR(Marshal.GetLastWin32Error());
+            Console.Write("Error DeviceIoControl " + Marshal.GetLastWin32Error());
+        }
+
+        FileStream file = File.Create(file_path);
+        file.Write(OutBufptr, 0, bytesReturned);
+        file.Close();
+
+        shwnd.Close();
```
Comment (tmolenh) [inline Get-NetView.psm1]: Probably should use try/finally to make sure these are closed.

File: Get-NetView.psm1 [hunk 2]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+        {
+            revision = IHV_DIAG_REVISION,
+            data_mask = data_mask_set
+        };
+
+        result = DeviceIoControl(shwnd, ioctl_value, ref check, Marshal.SizeOf(check), OutBufptr, bytesSize, ref bytesReturned, IntPtr.Zero);
+        if (false == result)
+        {
+            Exception e = Marshal.GetExceptionForHR(Marshal.GetLastWin32Error());
+            Console.Write("Error DeviceIoControl " + Marshal.GetLastWin32Error());
```
Comment (tmolenh) [inline Get-NetView.psm1]: How about writing the errors to the output file instead?

File: Get-NetView.psm1 [hunk 3]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+        }
+
+        FileStream file = File.Create(file_path);
+        file.Write(OutBufptr, 0, bytesReturned);
+        file.Close();
+
+        shwnd.Close();
+
+        if (result)
```
Comment (tmolenh) [inline Get-NetView.psm1]: Use braces `{}`

File: Get-NetView.psm1 [hunk 4]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+    }
+    else{
+        $MarvellChipCode = 1;
+        $MarvellVbdService = "qebdrv";
+        $MarvellNicService = "l2nd2";
+    }
+
+    # Collect Marvell related event logs and miscellaneous details
+    $file = "$NicName-Bus.txt"
+    [String []] $cmds = "Get-EventLog -Source $MarvellVbdService -LogName System | Format-List"
```
Comment (tmolenh) [inline Get-NetView.psm1]: Use `Get-WinEvent` instead for PowerShell Core compat.

File: Get-NetView.psm1 [hunk 5]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+    #"Get-EventLog -LogName System | where { $_.ProviderName -like $MarvellVbdService} | Format-List"
+    ExecCommandsAsync -OutDir $dir -File $file -Commands $cmds
+
+    $file = "$NicName-Nic.txt"
+    [String []] $cmds = "Get-EventLog -Source $MarvellNicService -LogName System | Format-List"
```
Comment (tmolenh) [inline Get-NetView.psm1]: Use `Get-WinEvent`

File: Get-NetView.psm1 [hunk 6]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+        }
+        else
+        {
+            data_mask_set = B10_IHV_COLLECT_MASK;
+            ioctl_value = B10_IOC_GET_IHV_DIAGNOSTIC_DATA;
+            local_path = local_path + "#{ea22615e-c443-434f-9e45-c4e32d83e97d}";
+        }
+
+        file_path = file_path + "\\";
+        file_path = file_path + filename;
```
Comment (tmolenh) [inline Get-NetView.psm1]: `file_path += "\\" + filename`

File: Get-NetView.psm1 [hunk 7]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
     MellanoxDetailPerNic -NicName $NicName -OutDir $dir
 } # MellanoxDetail()
 
+function MarvellDetail{
```
Comment (tmolenh) [inline Get-NetView.psm1]: If there are any Marvell specific counters you would like to collect, please add an appropriate filter to `$pathFilters` in function Counters

Comment (vikasmrvl) [reply Get-NetView.psm1]: Added Marvell counters. Thanks for highlighting it.

File: Get-NetView.psm1 [hunk 8]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)] [String] $NicName,
+        [parameter(Mandatory=$true)] [String] $OutDir
+    )
+
+$MarvellGetDiagDataClass = @"
```
[the remark this answers is not in the document]
Comment (vikasmrvl) [reply Get-NetView.psm1]: Verified in PowerShell Core.

File: Get-NetView.psm1 [hunk 9]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+
+    [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+    private static extern SafeFileHandle CreateFile(
+    string lpFileName,
+    [MarshalAs(UnmanagedType.U4)] FileAccess dwDesiredAccess,
+    [MarshalAs(UnmanagedType.U4)] FileShare dwShareMode,
+    IntPtr lpSecurityAttributes,
+    [MarshalAs(UnmanagedType.U4)] FileMode dwCreationDisposition,
+    [MarshalAs(UnmanagedType.U4)] FileAttributes dwFlagsAndAttributes,
+    IntPtr hTemplateFile);
```
[the remark this answers is not in the document]
Comment (vikasmrvl) [reply Get-NetView.psm1]: Added indentation.

File: Get-NetView.psm1 [hunk 10]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+    [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+    private static extern bool DeviceIoControl(
+    SafeFileHandle hDevice,
+    uint IoControlCode,
+    ref DiagInput_t InBuffer,
+    int nInBufferSize,
+    byte[] OutBuffer,
+    int nOutBufferSize,
+    ref int pBytesReturned,
+    IntPtr Overlapped
```
[the remark this answers is not in the document]
Comment (vikasmrvl) [reply Get-NetView.psm1]: Added indentation.

File: Get-NetView.psm1 [hunk 11]
```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
...
+        uint data_mask_set = 0;
+
+        if ((dev_path == null) || (file_path == null))
+        {
+            Console.Write("Input pointers are NULL \n");
+            return 0;
+        }
+
+        local_path = local_path.TrimEnd('\\');
+        if (!local_path.StartsWith(@"\\?\Global\"))
```
[the remark this answers is not in the document]
Comment (vikasmrvl) [reply Get-NetView.psm1]: Restructured code. This line no more exists.

File: Get-NetView.psm1 (hunk 1 above)
Comment (vikasmrvl) [reply Get-NetView.psm1]: Added try/catch/finally.

File: Get-NetView.psm1 (hunk 3 above)
Comment (vikasmrvl) [reply Get-NetView.psm1]: Done.

File: Get-NetView.psm1 (hunk 4 above)
Comment (vikasmrvl) [reply Get-NetView.psm1]: Done.

File: Get-NetView.psm1 (hunk 5 above)
Comment (vikasmrvl) [reply Get-NetView.psm1]: Done.

File: Get-NetView.psm1 (hunk 6 above)
Comment (vikasmrvl) [reply Get-NetView.psm1]: Done.

**Comment by vikasmrvl** [comment]
Thanks for the feedback. I have resolved all the comments. Please verify and let me know if there is 
more improvements required.

**Comment by tmolenh** [comment]
Perfect, Thanks!

**Pull request merged**